tests/physical_models/crop/test_leaf_dynamics.py

```diff
@@ -0,0 +1,238 @@
+
+        engine = EngineTestHelper(params, wdp, agro, config_path, external_states)
+        engine.run_till_terminate()
+        actual_results = engine.get_output()
+
+        # get expected results from YAML test data
+        expected_results, expected_precision = get_test_data(test_data_path)
+
+        assert len(actual_results) == len(expected_results)
+
+        for reference, model in zip(expected_results, actual_results, strict=False):
+            assert reference["DAY"] == model["day"]
+            assert all(
+                abs(reference[var] - model[var]) < precision
+                for var, precision in expected_precision.items()
+            )
+
+    def test_leaf_dynamics_with_engine(self):
+        # prepare model input
+        test_data_path = phy_data_folder / "test_leafdynamics_wofost72_01.yaml"
+        params, wdp, agro, _ = prepare_engine_input(test_data_path)
+
+        config_path = str(phy_data_folder / "WOFOST_Leaf_Dynamics.conf")
+
+        # Engine does not allows to specify `external_states`
+        with pytest.raises(ValueError):
+            Engine(params, wdp, agro, config_path)
+
+    def test_wofost_pp_with_leaf_dynamics(self):
+        # prepare model input
+        test_data_path = phy_data_folder / "test_potentialproduction_wofost72_01.yaml"
+        params, wdp, agro, _ = prepare_engine_input(test_data_path)
+
+        with patch(
+            'pcse.crop.leaf_dynamics.WOFOST_Leaf_Dynamics',
+            WOFOST_Leaf_Dynamics
+            ):
+            model = Wofost72_PP(params, wdp, agro)
+            model.run_till_terminate()
+            actual_results = model.get_output()
+
+        # get expected results from YAML test data
+        expected_results, expected_precision = get_test_data(test_data_path)
+
+        assert len(actual_results) == len(expected_results)
+
+        for reference, model in zip(expected_results, actual_results, strict=False):
+            assert reference["DAY"] == model["day"]
+            assert all(
+                abs(reference[var] - model[var]) < precision
+                for var, precision in expected_precision.items()
+            )
+
+
+class TestDiffLeafDynamicsTDWI:
+    def test_gradients_tdwi_lai_leaf_dynamics(self):
+        # prepare model input
+        test_data_path = phy_data_folder / "test_leafdynamics_wofost72_01.yaml"
+        params, wdp, agro, external_states = prepare_engine_input(test_data_path)
+        config_path = str(phy_data_folder / "WOFOST_Leaf_Dynamics.conf")
+
+        # create a model and optimizer
+        model = DiffLeafDynamics(params, wdp, agro, config_path, external_states)
+        tdwi = torch.nn.Parameter(torch.tensor(0.2, dtype=torch.float32))
+        output = model({"TDWI": tdwi})
+        lai = output[0, :, 0]
+        loss = lai.sum()
+        grads = torch.autograd.grad(loss, tdwi, retain_graph=True)[0]  # this is ∂loss/∂tdwi
+
+        assert grads is not None, "Gradients for TDWI should not be None"
```

Comment:
Maybe a bit overkill for now but I recently learned about subtasks. https://pypi.org/project/pytest-subtests/
This would the code to continue running even after failing a single assert so all checks will be run always.
Something to try later maybe.

Reply:
nice suggestion! for now I skip adding this library as a dependency. I will add it as an issue to look at it later.